pkg/env/env.go

```diff
@@ -160,70 +172,134 @@ func (e *testEnv) AfterEachTest(funcs ...types.TestEnvFunc) types.Environment {
 	return e
 }
 
-// Test executes a feature test from within a TestXXX function.
-//
-// Feature setups and teardowns are executed at the same *testing.T
-// contextual level as the "test" that invoked this method. Assessments
-// are executed as a subtests of the feature.  This approach allows
-// features/assessments to be filtered using go test -run flag.
-//
-// Feature tests will have access to and able to update the context
-// passed to it.
-//
-// BeforeTest and AfterTest operations are executed before and after
-// the feature is tested respectively.
-func (e *testEnv) Test(t *testing.T, testFeatures ...types.Feature) {
+// panicOnMissingContext is used to check if the test Env has the right context setup
+// and fail fast if the context has not already been set
+func (e *testEnv) panicOnMissingContext() {
 	if e.ctx == nil {
 		panic("context not set") // something is terribly wrong.
 	}
+}
 
-	if len(testFeatures) == 0 {
-		t.Log("No test testFeatures provided, skipping test")
-		return
-	}
-
-	// execute the beforeTest functions
-	beforeTestActions := e.getBeforeTestActions()
+// processTestActions is used to run a series of test action that were configured as
+// BeforeEachTest or AfterEachTest
+func (e *testEnv) processTestActions(t *testing.T, actions []action) {
 	var err error
-	for _, action := range beforeTestActions {
+	for _, action := range actions {
 		if e.ctx, err = action.runWithT(e.ctx, e.cfg, t); err != nil {
 			t.Fatalf("BeforeEachTest failure: %s", err)
 		}
 	}
+}
+
+// processTestFeature is used to trigger the execution of the actual feature. This function wraps the entire
+// workflow of orchestrating the feature execution be running the action configured by BeforeEachFeature /
+// AfterEachFeature. This can be invoked with a nillable wg which defines how the test feature is being run.
+// If the wg is set to nil that means the test is being run in sequence and there is no reason to mark a
+// WaitGroup as done
+func (e *testEnv) processTestFeature(wg *sync.WaitGroup, t *testing.T, featureName string, feature types.Feature) {
```

Comment:
I don't think this should take a workgroup. I suggest to keep generic as possible for clarity (see how in next comment).